store/sqlstore/channel_store.go

```diff
@@ -1814,6 +1815,70 @@ func (s SqlChannelStore) UpdateLastViewedAt(channelIds []string, userId string)
 	return times, nil
 }
 
+// CountUnreadSince gives the number of posts that a channel has since some date, this doesn't follow any threads.
+// this allows us to know how many posts have bean read already.
+func (s SqlChannelStore) CountUnreadSince(channelID string, since int64) (int64, *model.AppError) {
+	countUnreadQuery := `
+	SELECT count(*)
+	FROM Posts
+	WHERE
+		ChannelId = :channelId
+		AND CreateAt > :createAt
+		AND Type = ''
+		AND DeleteAt = 0
+	`
+	countParams := map[string]interface{}{
+		"channelId": channelID,
+		"createAt":  since,
+	}
+
+	unread, err := s.GetReplica().SelectInt(countUnreadQuery, countParams)
+	if err != nil {
+		return 0, model.NewAppError("SqlChannelStore.CountUnreadSince", "store.sql_channel.CountUnreadSince.app_error", countParams, fmt.Sprintf("Error while counting unreads for %s since %d", channelID, since), http.StatusInternalServerError)
+	}
+	return unread, nil
+}
+
+// UpdateLastViewedAtPost sets a channel as unread for a user at the time of the post selected and update the MentionCount
+// it returns a channelunread so redux can update the apps easily.
+func (s SqlChannelStore) UpdateLastViewedAtPost(unreadPost model.Post, userID string, mentionCount int) (*model.ChannelUnread, *model.AppError) {
+
+	unread, appErr := s.CountUnreadSince(unreadPost.ChannelId, unreadPost.CreateAt)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	params := map[string]interface{}{
+		"mentions":     mentionCount,
+		"unreadCount":  unread,
+		"lastViewedAt": unreadPost.CreateAt,
+		"userId":       userID,
+		"channelId":    unreadPost.ChannelId,
+		"updatedAt":    utils.MillisFromTime(time.Now()),
+	}
+
+	// msg count uses the value from channels to prevent counting on older channels where no. of messages can be high.
+	// we only count the unread which will be a lot less in 99% cases
+	setUnreadQuery := `
+	UPDATE
+		ChannelMembers
+	SET
+		MentionCount = :mentions,
+		MsgCount = (SELECT TotalMsgCount FROM Channels WHERE ID = :channelId) - :unreadCount,
+		LastViewedAt = :lastViewedAt,
+		LastUpdateAt = :updatedAt
+	WHERE
+		UserId = :userId
+		AND ChannelId = :channelId
+	`
+	_, err := s.GetMaster().Exec(setUnreadQuery, params)
+	if err != nil {
+		return nil, model.NewAppError("SqlChannelStore.UpdateLastViewedAtPost", "store.sql_channel.UpdateLastViewedAtPost.app_error", params, "Error setting channel "+unreadPost.ChannelId+" as unread: "+err.Error(), http.StatusInternalServerError)
+	}
+
+	return s.GetChannelUnread(unreadPost.ChannelId, userID)
```

Comment:
I hadn't thought about this before, but this might cause issues due to lag between multiple database replicas. We update the master database above, but then we get the `ChannelUnread` back from a read replica which may not have up to date information about the number of unread messages yet.

I'm not sure how to fix this, so unless you have any ideas for how to fix it, maybe we can start a discussion about it on community since other people might have ideas.

Reply:
I could definitely duplicate the code using master, but i usually dislike that unless justified to prevent modifying one and forgetting about the other. We could either have some common source having the database to use as a parameter or just throw a comment in both places to ensure someone modifying one looks at the other.

OTOH I'm wondering if we are trying to solve a problem we don't have. At worst, there will be a delay informing the user as redux will get the wrong info, but there is a process to update that later on (I've tested this change without redux and it appears in a few seconds), with redux and websockets this should be almost instant if replica gets it in time.

Reply:
Yeah, it's possible this might not be an issue. Since there's no obvious fix for it, I'll just make a note of it for when we're testing with HA. The worst thing that could happen anyway is that the feature might not work when the server is under a heavy load, but it wouldn't break anything else or cause any lasting issues.